src/org/armedbear/lisp/Lisp.java

```diff
@@ -517,6 +583,73 @@ public static final LispObject eval(LispObject obj)
+      LispObject functionSetStepCounterCompleted = coerceToFunction(symbolSetStepCounterCompleted);
+      LispObject[] values = currentThread._values;
+      funcall(functionSetStepCounterCompleted,
+              new LispObject[] {LispInteger.getInstance(stepNumberInternal)},
+              currentThread);
+      currentThread._values = values;
+    }
+  }
+
+  public static synchronized final void handleStepping (LispObject function, LispObject args,
+                                                        Environment env, LispInteger stepCount) {
+    LispThread currentThread = LispThread.currentThread();
+    Package stepper = Packages.findPackageGlobally("ABCL-STEPPER");
+    Package system = Packages.findPackageGlobally("SYSTEM");
+    Symbol symbolPprintFormToStep = stepper.findAccessibleSymbol("PPRINT-FORM-TO-STEP");
+    Symbol symbolHandleUserInteraction = stepper.findAccessibleSymbol("HANDLE-USER-INTERACTION");
+    LispObject functionPprintFormToStep = coerceToFunction(symbolPprintFormToStep);
+    LispObject functionHandleUserInteraction = coerceToFunction(symbolHandleUserInteraction);
+    if (stepperBlock == null) {
+      stepperBlock = env.getOuterMostBlock();
+    }
+    if (function instanceof FuncallableStandardObject) {
+      function = ((FuncallableStandardObject)function).function;
+    }
+    LispObject closureName = ((Operator)function).getLambdaName();
+    setSteppingOff();
+    if (closureName != null ) {
+      funcall(functionPprintFormToStep, new LispObject[] {closureName, args, stepCount}, currentThread);
+    }
+    else {
+      funcall(functionPprintFormToStep, new LispObject[] {((Operator)function), args, stepCount}, currentThread);
+    }
+    setSteppingOn();
+    funcall(functionHandleUserInteraction, new LispObject[]{env}, currentThread);
+  }
+
+  public static final void printStepValue(long stepNumberInternal, LispObject result, LispThread thread) {
+    Package stepper = Packages.findPackageGlobally("ABCL-STEPPER");
+    Package system = Packages.findPackageGlobally("SYSTEM");
+    Symbol symbolPrintStepperStr = stepper.findAccessibleSymbol("PRINT-STEPPER-STR");
+    LispObject functionPrintStepperStr = coerceToFunction(symbolPrintStepperStr);
+    LispObject[] values = thread._values;
+    if (values != null) {
+      for (int i = 0; i < values.length; i++) {
+        funcall(functionPrintStepperStr,
+                new LispObject[] {
+                  new SimpleString("step " + stepNumberInternal + " ==> value: " + values[i].printObject()),
+                  Symbol.T
+                },
+                thread);
+      }
+    }
+    else {
+      funcall(functionPrintStepperStr, new LispObject[] {
+          new SimpleString("step " + stepNumberInternal + " ==> value: " + result.printObject()),
+          Symbol.T
+        },
+        thread);
+    }
+    thread._values = values;
+  }
+
   public static final LispObject eval(final LispObject obj,
```

Comment:
Have you done any performance benchmarks?  We are adding additional steps to the Lisp.eval() codepath here, even if we aren't stepping through functions.  From looking at this code, it seems like the only "penalty" is in the boolean compares for `stepInSymbolResult` et. al. but it would be nice to confirm there isn't a generic speed penalty here.